### PR TITLE
[DataGrid] Handle special chars when doing CSV export

### LINKED ISSUE #1143: [DataGrid] CSV export incorrect if cell data contains comma
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯
image
image

## Expected Behavior 🤔
image

## Steps to Reproduce 🕹

Steps:

1.open https://codesandbox.io/s/material-demo-forked-zdvow?file=/demo.tsx
2.click `Export` -> `Download as CSV`

### PR DESCRIPTION
**Pull request by DanailH**
Fixes #1143

### CODE REVIEW

**Review by oliviertassinari — CHANGES_REQUESTED** [review]
- How about we add a test case that reproduce the issue?
- What happens if there a double quote in the cell we are exporting? See http://www.creativyst.com/Doc/Articles/CSV/CSV01.htm#FileFormat
- Should we only add the " if there is a comma in order to save space in the CSV export?

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts [hunk 1]
```diff
@@ -1,8 +1,10 @@
 import { GridColumns, GridRowId, GridRowModel } from '../../../../models';
 
-export function buildRow(row: GridRowModel, columns: GridColumns): Array<GridRowModel> {
-  const mappedRow: GridRowModel[] = [];
-  columns.forEach((column) => column.field !== '__check__' && mappedRow.push(row[column.field]));
+export function buildRow(row: GridRowModel, columns: GridColumns): Array<string> {
+  const mappedRow: string[] = [];
+  columns.forEach(
+    (column) => column.field !== '__check__' && mappedRow.push(`"${row[column.field]}"`),
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: this doesn't care about valueGetter in colDef.
I'm adding a `getCellValue` in my editRows PR. But not exposed so maybe we should expose it.

Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: BTW you only need to wrap it with "" when there are special chars

Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: Ok, I think we need to extend the functionality a bit. I spent some time reading the specks yesterday and check how others solve that issue and here is the outcome. The problem is more complex than it seems.

Most competitors have a default functionality that works OTB with the simplest case (in our example that would be to either keep it as it is or wrap the cell values in `"`). I think that wrapping it with `"` will work for most cases. In addition they have a way for a developer to pass in their delimiter/separator to be used for formatting, meaning that if the default is `"` then if you have values with `"` in it then you can choose to use `,`. That way if the OTB functionality doesn't work for a developer's use case they can implement them themselves by calling the provided API and pass in their delimiter/separator.

On the topic of handling that case, ourselves -> I played a bit with a potential algorithm but the issue is that you need to traverse the entire data at least 2 times to make it work, which is a case of very big data will be time-consuming. In the basic case, the complexity is O(2n) but the problem is that if you have `"` and `,` and `|` then it wouldn't be possible to format it, basically, I don't think we can handle all the cases ourselves.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: Could you share the links of what you have explored so far? 

This solution seems to add double quotes for comma and escape double quotes. Would it be enough? If not, why? https://stackoverflow.com/a/24922761 not enough?

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: Maybe we could salvage logic from https://github.com/jmaister/excellentexport/blob/master/src/excellentexport.ts

Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: I think it won't cover all the cases. The issue is that you have 3 ways to format a csv `"` or `,` or `|`, and if the data you want to format has all 3 then how would you decide which to use?

In terms of links -> https://www.ag-grid.com/documentation/vue/csv-export/ for example.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: What do you mean by "format"?

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: Material Table is using https://github.com/mbrn/filefy/blob/master/src/csv-builder/CsvBuilder.ts

Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: The current way the CSV works is that it only supports `,` as a delimiter. I think having the option for developers to set their own delimiters through the CSV export API will work because I don't think we can handle all the cases where the data contains different potential delimiters?

Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:6]: I agree we should only start by supporting ',' as it's the most common use case. Then wrap with "" and escape the value as in `escapeCell ` or `fixCSVField ` of the link shared above.

Using the proper value is more important.

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts [hunk 2]
```diff
@@ -1,9 +1,16 @@
 import { GridColumns, GridRowId, GridRowModel } from '../../../../models';
 
+const formatCellValue = (value) => {
+  if (typeof value === 'string') {
+    return value.includes(',') ? `"${value}"` : value;
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:5]: what if the value contains dbl quote?

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts (hunk 1 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts]: Ok, I've updated the formatter. It will now handle the case where the delimiter is `,` and if there is a `,` in a cell value that cell value will be wrapped in a `"` resulting in a correctly formatted CSV. I also added an addition to the existing test case.

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts (hunk 2 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:5]: it will still work as expected, the problem comes when you have a structure like this `value"value,value"value`

Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:5]: but if you don't escape the double quote it will be parsed as another col if it contains the ','

ie cell value: 'abc , "quote, from a great guy" is easy '

Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:5]: I will a do PR that will add a getCellValue on apiref

Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts]: The case is now handled

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts (hunk 1 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts]: It works with the latest changes

File: packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx [hunk 3]
```diff
@@ -154,7 +154,13 @@ describe('<XGrid /> - apiRef', () => {
 
   it('getDataAsCsv should return the correct string representation of the grid data', () => {
     render(<TestCase />);
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx:156]: I think that we should refactor the least possible in the tests. DRY is our enemy. How about we render just the data we need instead of calling `updateRows`?

**Review by oliviertassinari — CHANGES_REQUESTED** [review]
Regarding the tests, I think that we can improve them:

1. Only test the CSV exporter without the data grid. We will need to so more or less the same for the Excel one
2. Cover edge cases you have considered but aren't tested. e.g. `"` inside the value of one cell.

File: packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx (hunk 3 above)
Comment (DanailH) [reply packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx:156]: No preference. I followed the approach done for the other tests and the `TestCase` does work for that use case.

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts [hunk 4]
```diff
@@ -1,8 +1,18 @@
 import { GridColumns, GridRowId, GridRowModel } from '../../../../models';
 
-export function buildRow(row: GridRowModel, columns: GridColumns): Array<GridRowModel> {
-  const mappedRow: GridRowModel[] = [];
-  columns.forEach((column) => column.field !== '__check__' && mappedRow.push(row[column.field]));
+const formatCellValue = (value) => {
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:3]: we could name these 2 func, `serialiseRow, serialiseCell`

File: packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx [hunk 5]
```diff
@@ -154,7 +154,13 @@ describe('<XGrid /> - apiRef', () => {
...
-
     expect(apiRef.current.getDataAsCsv()).to.equal('Brand\r\nNike\r\nAdidas\r\nPuma');
+    apiRef.current.updateRows([
+      {
+        id: 1,
+        brand: 'Adidas,Reebok',
```
Comment (dtassone) [inline packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx:161]: could we had a test with `brand: 'A "d,i,das '`

File: packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx (hunk 3 above)
Comment (oliviertassinari) [reply packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx:156]: I think that providing props for the rows would already be better than using the `apiRef.current.updateRows` API. The more declartive we can get (avoiding imperative calls) and multiple renders, the better.

File: packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx (hunk 5 above)
Comment (DanailH) [reply packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx:198]: done

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts (hunk 4 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts]: done

File: packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx (hunk 3 above)
Comment (DanailH) [reply packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx]: Done, I split the test case and added 2 more.

**Review by oliviertassinari — APPROVED** [review]
Things I would have done differently if I was working on this:

- In the tests, I would have used 2-dimensional data. Note one row, nor one column, but at least two columns and rows.
- I would have split the problem into two phases. One phase about constructing the 2d data structure for export. And a second phase to convert this 2d structure into a CSV. Hopefully, this would bring two values. I suspect that the first phase would be shared with the Excel export. I also assume that it would allow, in the future to write unit tests that run much faster for covering more formatting edge-cases and more formating options, like custom delimiter.

File: packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx (hunk 3 above)
Comment (oliviertassinari) [reply packages/grid/x-grid/src/tests/apiRef.XGrid.test.tsx]: Thanks. Speaking of abstraction and duplication in the tests. I suspect that we could abstract the usage of the apiRef as well as the container dimensions. We will see.

File: packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts [hunk 6]
```diff
@@ -1,15 +1,32 @@
...
+
+export function serialiseRow(
+  row: GridRowModel,
+  columns: GridColumns,
+  getCellValue: (id: GridRowId, field: string) => GridCellValue,
+): Array<string> {
+  const mappedRow: string[] = [];
+  columns.forEach(
+    (column) =>
+      column.field !== '__check__' &&
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts:19]: This one `'__check__'` should be a constant

Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/export/seralizers/csvSeraliser.ts]: Done

**Review by dtassone — APPROVED** [review]

**Comment by z0d14c** [comment]
I started using mui x datagrid well after this was merged (end of 2023) and I still see the issue... I'm just curious what is the best way to check if the commit is in the version of mui x I'm using?

**Comment by z0d14c** [comment]
On `v6.16.0`, I'm not seeing anything when I search `git log --all --grep="82f3d84"` ...

**Comment by z0d14c** [comment]
hmm, nevermind, I found it by searching for the content of the commit:

commit 82f3d8485852b2fabfa9592b25d850d966fb8705
Author: Danail Hadjiatanasov <hadjiatanasov@gmail.com>
Date:   Mon Mar 8 17:14:56 2021 +0100

    [DataGrid] Handle special chars when doing CSV export (#1154)
    
wonder why it isn't working -- we're still seeing screwed up exports when we have values with commas in them
